sycl/include/sycl/ext/oneapi/experimental/complex/detail/marray.hpp

```diff
@@ -0,0 +1,316 @@
+      lhs[i] op rhs;                                                           \
+    }                                                                          \
+    return lhs;                                                                \
+  }                                                                            \
+  friend marray &operator op(value_type & lhs, const marray & rhs) {           \
+    for (std::size_t i = 0; i < NumElements; ++i) {                            \
+      lhs[i] op rhs;                                                           \
+    }                                                                          \
+    return lhs;                                                                \
+  }
+
+  MARRAY_CPLX_OP(+=)
+  MARRAY_CPLX_OP(-=)
+  MARRAY_CPLX_OP(*=)
+  MARRAY_CPLX_OP(/=)
+
+#undef MARRAY_CPLX_OP
+
+  // MARRAY_CPLX_OP is: %=
+  friend marray &operator%=(marray &lhs, const marray &rhs) = delete;
+  friend marray &operator%=(marray &lhs, const value_type &rhs) = delete;
+  friend marray &operator%=(value_type &lhs, const marray &rhs) = delete;
+
+// MARRAY_CPLX_OP is: ++, --
+#define MARRAY_CPLX_OP(op)                                                     \
+  friend marray operator op(marray &lhs, int) = delete;                        \
+  friend marray &operator op(marray & rhs) = delete;
+
+  MARRAY_CPLX_OP(++)
+  MARRAY_CPLX_OP(--)
+
+#undef MARRAY_CPLX_OP
+
+// MARRAY_CPLX_OP is: unary +, unary -
+#define MARRAY_CPLX_OP(op)                                                     \
+  friend marray<value_type, NumElements> operator op(                          \
+      const marray<value_type, NumElements> &rhs) {                            \
+    marray<value_type, NumElements> rtn;                                       \
+    for (std::size_t i = 0; i < NumElements; ++i) {                            \
+      rtn[i] = op rhs[i];                                                      \
+    }                                                                          \
+    return rtn;                                                                \
+  }
+
+  MARRAY_CPLX_OP(+)
+  MARRAY_CPLX_OP(-)
+
+#undef MARRAY_CPLX_OP
+
+// MARRAY_CPLX_OP is: &, |, ^
+#define MARRAY_CPLX_OP(op)                                                     \
+  friend marray operator op(const marray &lhs, const marray &rhs) = delete;    \
+  friend marray operator op(const marray &lhs, const value_type &rhs) = delete;
+
+  MARRAY_CPLX_OP(&)
+  MARRAY_CPLX_OP(|)
+  MARRAY_CPLX_OP(^)
+
+#undef MARRAY_CPLX_OP
+
+// MARRAY_CPLX_OP is: &=, |=, ^=
+#define MARRAY_CPLX_OP(op)                                                     \
+  friend marray &operator op(marray & lhs, const marray & rhs) = delete;       \
+  friend marray &operator op(marray & lhs, const value_type & rhs) = delete;   \
+  friend marray &operator op(value_type & lhs, const marray & rhs) = delete;
+
+  MARRAY_CPLX_OP(&=)
+  MARRAY_CPLX_OP(|=)
+  MARRAY_CPLX_OP(^=)
```

Comment:
Why not have `%=` deleted here too?

Reply:
The reason for not having the `%=` operator being deleted with the `bitwise` assignment operator is because I arranged the operator by "type". The `modulo` operator not being a `bitwise`  operation felt like it had to be on its own.

- Arithmetic
- Arithmetic assignment
- Bitwise
- Bitwise assignment
- etc ...